### ISSUE TITLE
test folder

### ISSUE THREAD

**Issue by leedz0210** [dodaltuin/soft-tissue-pignn#4]
Hi author, thank you for a great job that I benefited from, it was a great job. But I am a newbie and don't know how to get pe-values.npy, displacement.npy, theta.npy in the test folder, please can you upload such an example with a view to learn

**Comment by dodaltuin** [comment]
Hi there, yes that is no problem - please see the notebook [in this directory](https://github.com/dodaltuin/soft-tissue-pignn/tree/main/data/fenicsSimulationCode) for details on how to compute and save the pe-values.npy, displacement.npy and theta.npy files.

**Comment by vic663** [comment]
Hello Author,

Thank you for your amazing work on this repository! I have learned a lot from your implementation, and I really appreciate your contributions.

I am currently working on the Left Ventricle (LV) simulation using the Holzapfel-Ogden material model, and I saw your previous response regarding the pe-values.npy, displacement.npy and theta.npyfiles, but I am still a bit confused about how to generate them correctly for the Left Ventricle (LV) case.

Would you be able to upload an example script or notebook showing how to compute and save these files specifically for the LV simulation?

This would be incredibly helpful for me and others trying to understand how to extract meaningful outputs from the simulation.

Thanks in advance for your help!

**Comment by dodaltuin** [comment]
Hi there,

Yeah things are a bit more tricky with the LV model - I'm actually having a meeting tomorrow with my coauthor @HaoGao, where I hope to discuss some of these issues relating to the FEniCS code. I will get back to you with more details after that.

**Comment by vic663** [comment]
Thank you for your response! I’ll be happy to wait for your updates after the meeting. Looking forward to your insights!

**Comment by leedz0210** [comment]
Thank you for your reply, it is appreciated, I will try to generate the data, thanks again!

**Comment by dodaltuin** [comment]
Hi @vic663, having had a chat with @HaoGao, my plan is to get a notebook for the LV simulation ready and available here next week. I'm very busy at the minute with teaching, however, so it may be two weeks before I have time to do this.

**Comment by vic663** [comment]
Hi @dodaltuin , thanks for the update! I completely understand how busy things can get, especially with teaching commitments. In the meantime, if you have any quick hints, tips, or resources I could look into to get started, that would be incredibly helpful!

**Comment by dodaltuin** [comment]
You might find the notebook [here](https://github.com/ComputationalPhysiology/Suurph_mechanics_docker/blob/master/L21A-solved.ipynb) useful - we based our LV simulation code off of this.

**Comment by vic663** [comment]
Thank you for sharing the notebook reference!

**Comment by vic663** [comment]
Hi @dodaltuin , thanks for sharing the notebook! I noticed that the LV code references theta-mean.npy and theta-std.npy. Could you provide some guidance on how these files were generated? Are they derived from a specific dataset or preprocessing step? Any details would be super helpful!

**Comment by dodaltuin** [comment]
There are basically two ways of calculating these values:

1. If you have an $(N \times D_{\theta})$ numpy array ``theta_values`` which contains all the parameter values you want to work with, then  ``theta_mean`` and ``theta_std`` are just the column-wise mean and standard deviation respectively of this array. This can be computed with the following code:

	```python
	theta_mean = theta_values.mean(0)
	theta_std  = theta_values.std(0)
	```

2. If you have a population probability distribution over parameter space, then ``theta_mean`` and ``theta_std`` are just the mean and standard deviation of this distribution. So, for instance, if you believe that material parameter values are uniformly distributed in a "box" in $\mathbb{R}^{D_{\theta}}$, then you can evaluate the mean and standard deviation using the well known formulas [here](https://en.wikipedia.org/wiki/Continuous_uniform_distribution).

**Comment by vic663** [comment]
Thank you for the detailed explanation!

**Comment by vic663** [comment]
I have a follow-up question: Could you please clarify how to obtain or generate the files fibre_dir.xml, lvgmsh_fmaker.xml, and lvgmsh_mesh.xml? Are these files typically created using specific software tools, or are they part of a predefined dataset or library? If they need to be generated, could you provide some guidance on the scripts or tools required to create them?

**Comment by HaoGao** [comment]
It involves quite a few steps to prepare those files, you will need to be familiar with Fencics. The steps I have followed (1) a geometry file in a format Gmsh can read it and define surfaces; (2) generate mesh and output using gmsh; (3) convert the gmsh file into fenics format, the XML file format. this is the code I used https://gist.github.com/finsberg/2a919ad0beb0f4f169294ce34bea6fed; (4) using the following library to define the fibre architecture, https://github.com/finsberg/ldrb. After that, you should have all the required files.  You may come across some incompatibility issues since some of the codes are quite old. I am planning to include an example in the future, probably over the summer when I have some bit of time.

**Comment by dodaltuin** [comment]
Hi @vic663  I have now added  this notebook [here](https://github.com/dodaltuin/soft-tissue-pignn/blob/main/data/fenicsSimulationCode/LeftVentricle.ipynb), which contains FEniCS simulation code for the ``LeftVentricle`` example. While the simulation does run, it is however quite slow.

**Comment by vic663** [comment]
Hi @dodaltuin Thanks for sharing this! I’ll take a look.

**Comment by vic663** [comment]
Hi @dodaltuin I was exploring this notebook [here](https://github.com/dodaltuin/soft-tissue-pignn/blob/main/data/fenicsSimulationCode/LeftVentricle.ipynb) and noticed the infrastructure for load steps. However, this doesn’t seem to be actively used in the main training loop.

Could you clarify: Was progressive loading intended to improve stability for quasi-static problems? If so, is there a recommended way to integrate it into the training loop?

Training Stability: When applying the framework to new LV geometries, I'm encountering stability issues (e.g., NaN gradients, poor convergence). Are there recommended strategies in the existing framework to improve stability? For example:

Specific parameter initializations

Modified loss weighting

Architectural adjustments for new geometries

Would you suggest any particular approach for maintaining stability when extending to new geometries? Any insights would be greatly appreciated.

Best regards,
